Change containerSecurityContext rendering and add docs

janlauber opened the pull request:
Facing the discussion of: 
- #6686 
- #6685 

It adds the functionality of adding `securityContext: null` to your custom values.yaml in case of deploying the dashboard deployment without the securityContext inside your cluster. 
When this line is not added to your custom values.yaml file it will render the securityContext values as defined in the default values.yaml file.

Signed-off-by: Jan Lauber <jan.lauber@protonmail.ch>

desaintmartin commented:
I am not sure what the changes introduces.

"with" is true according to the same rules as "if". So how a falsy value (`null` in your case) would be interpreted as true with `if` better than `with`?

A better way would be to state it unconditionally with sane default value.

desaintmartin commented:
/assign

floreks commented:
I still think that the issue is inside their cluster and not our helm chart. I'll test tomorrow on a clean v1.21 k8s cluster to confirm. This change does not actually change anything for us as @desaintmartin said.

floreks commented:
Hold until I test this on a clean cluster.

/hold

janlauber commented:
So I've tested it again with `securityContext: null` locally and it works.
As @desaintmartin already said, the if / with wouldn't change anything.
So I reverted my last commit and added just one line of docs (comment) in the default values.yaml file so if anybody in the future also stumbles over it.

Thanks for your work!
greez

desaintmartin commented:
OK Understood, your comment is clear and not obvious (as you actually need to ADD a line in your custom values to unset a parameter in the default values).

Could you bump the patch version (according to semver) in chart.yaml?

janlauber commented:
@desaintmartin 
I bumped the version to 5.1.1 and the CLA check should be passed.
Thanks and greez

desaintmartin commented:
One last thing: could you edit PR name?

janlauber commented:
sth like that?

desaintmartin commented:
Yes, thanks!
/lgtm

janlauber commented:
probably `do-not-merge/hold` label should get removed to merge

desaintmartin commented:
I let @floreks remove the lock he put if he agrees. :)

floreks commented:
## Environment
```
Kubernetes version: v1.21.1
Cluster deployment: Kind
Helm version: v3.7.0
```

## Installation 
```
helm install kubernetes-dashboard kubernetes-dashboard/kubernetes-dashboard -f values.yaml
```

## Default Values
[values.yaml](https://github.com/kubernetes/dashboard/blob/master/aio/deploy/helm-chart/kubernetes-dashboard/values.yaml)

## Scenario 1 (default values)

### Result
Deployment spec contains the below security context
```yaml
spec:
  securityContext:
    seccompProfile:
      type: RuntimeDefault
  containers:
    securityContext:
        allowPrivilegeEscalation: false
        readOnlyRootFilesystem: true
        runAsGroup: 2001
        runAsUser: 1001
```

Pod spec contains the below security context
```yaml
spec:
  securityContext:
    seccompProfile:
      type: RuntimeDefault
  containers:
    securityContext:
        allowPrivilegeEscalation: false
        readOnlyRootFilesystem: true
        runAsGroup: 2001
        runAsUser: 1001
```

## Scenario 2 (security context null)
In this scenario, we'll set the `securityContext` to `null` in `values.yaml` file (line 74).
```yaml
securityContext: null
```

### Result
Deployment spec contains the below security context
```yaml
spec:
  containers:
    securityContext:
        allowPrivilegeEscalation: false
        readOnlyRootFilesystem: true
        runAsGroup: 2001
        runAsUser: 1001
```

Pod spec contains the below security context
```yaml
spec:
  containers:
    securityContext:
        allowPrivilegeEscalation: false
        readOnlyRootFilesystem: true
        runAsGroup: 2001
        runAsUser: 1001
```

### Scenario 3 (security context null + container security context null)
In this scenario, we'll also set the `containerSecurityContext` to `null` in `values.yaml` file (line 313).

### Result
Both deployment spec and pod spec do not contain any security context.

floreks commented:
Unfortunately, it appears that security context configuration, in general, is scattered around the `values.yaml` file and it is not obvious how to control it properly. We should have one simple way of controlling security context configuration across all our deployments. 

General context added on the `spec` level.
https://github.com/kubernetes/dashboard/blob/1148f7ba9f9eadd719e53fa3bc8bde5b7cfdb395/aio/deploy/helm-chart/kubernetes-dashboard/values.yaml#L73-L76

Metrics scraper context added on the `spec.containers` level.
https://github.com/kubernetes/dashboard/blob/1148f7ba9f9eadd719e53fa3bc8bde5b7cfdb395/aio/deploy/helm-chart/kubernetes-dashboard/values.yaml#L239-L244

Dashboard context added on the `spec.containers` level.
https://github.com/kubernetes/dashboard/blob/1148f7ba9f9eadd719e53fa3bc8bde5b7cfdb395/aio/deploy/helm-chart/kubernetes-dashboard/values.yaml#L312-L317

@janlauber @desaintmartin it is actually a mistake on our side as it is indeed misleading how to control this configuration. It should be refactored and simplified.

desaintmartin commented:
Indeed, at least what we can do is put the `securityContext` and containerSecurityContext next to each other!

Regarding `metricsScraper`, I do not think it would be wise to put a parameter tied to this container outside of this top-level value.

floreks commented:
Couldn't we just use a single configuration for everything? We anyway want to use the same security context for both dashboard and scraper.

janlauber commented:
@floreks I think this would make sense to use a single configuration for both.
I can start trying to implement this in this PR if it's ok?
greez

floreks commented:
I think you can start working on that if you want :slightly_smiling_face:

janlauber commented:
I would like to try something like in the kube-prometheus-stack helm chart:
https://github.com/prometheus-community/helm-charts/blob/c607b596cd6bc49a02b8e142108a91d1089fb781/charts/kube-prometheus-stack/values.yaml#L101-L105

Adding a `global` section to set some values globally.
Would this also make sense to you? @floreks

janlauber commented:
@desaintmartin & @floreks 
I've just changed the behaviour of rendering the containerSecurityContext on a single values next to the securityContext.
Tests cases:
Helm template test command in `dashboard/aio/deploy/helm-chart/kubernetes-dashboard`
```bash
helm template -f values-test.yaml --dry-run .
```
**1. default values in custom values-test.yaml:**

```yaml
# empty
```
Output of the Deployment:
```yaml
apiVersion: apps/v1
kind: Deployment
metadata:
  name: RELEASE-NAME-kubernetes-dashboard
  annotations:
  labels:
    app.kubernetes.io/name: kubernetes-dashboard
    helm.sh/chart: kubernetes-dashboard-5.1.1
    app.kubernetes.io/instance: RELEASE-NAME
    app.kubernetes.io/version: "2.4.0"
    app.kubernetes.io/managed-by: Helm
    app.kubernetes.io/component: kubernetes-dashboard
spec:
  replicas: 1
  strategy:
    rollingUpdate:
      maxSurge: 0
      maxUnavailable: 1
    type: RollingUpdate
  selector:
    matchLabels:

      app.kubernetes.io/name: kubernetes-dashboard
      app.kubernetes.io/instance: RELEASE-NAME
      app.kubernetes.io/component: kubernetes-dashboard
  template:
    metadata:
      annotations:
      labels:
        app.kubernetes.io/name: kubernetes-dashboard
        helm.sh/chart: kubernetes-dashboard-5.1.1
        app.kubernetes.io/instance: RELEASE-NAME
        app.kubernetes.io/version: "2.4.0"
        app.kubernetes.io/managed-by: Helm
        app.kubernetes.io/component: kubernetes-dashboard
    spec:
      securityContext:

        seccompProfile:
          type: RuntimeDefault
      serviceAccountName: RELEASE-NAME-kubernetes-dashboard
      containers:
      - name: kubernetes-dashboard
        image: "kubernetesui/dashboard:v2.4.0"
        imagePullPolicy: IfNotPresent
        args:
          - --namespace=default
          - --auto-generate-certificates
          - --metrics-provider=none
        ports:
        - name: https
          containerPort: 8443
          protocol: TCP
        volumeMounts:
        - name: kubernetes-dashboard-certs
          mountPath: /certs
          # Create on-disk volume to store exec logs
        - mountPath: /tmp
          name: tmp-volume
        livenessProbe:
          httpGet:
            scheme: HTTPS
            path: /
            port: 8443
          initialDelaySeconds: 30
          timeoutSeconds: 30
        resources:

          limits:
            cpu: 2
            memory: 200Mi
          requests:
            cpu: 100m
            memory: 200Mi
        securityContext:

          allowPrivilegeEscalation: false
          readOnlyRootFilesystem: true
          runAsGroup: 2001
          runAsUser: 1001
      volumes:
      - name: kubernetes-dashboard-certs
        secret:
          secretName: RELEASE-NAME-kubernetes-dashboard-certs
      - name: tmp-volume
        emptyDir: {}
```
**2. default values in custom values-test.yaml with metrics scraper enabled:**
```yaml
metricsScraper:
  enabled: true
```
Output of the Deployment:
```yaml
apiVersion: apps/v1
kind: Deployment
metadata:
  name: RELEASE-NAME-kubernetes-dashboard
  annotations:
  labels:
    app.kubernetes.io/name: kubernetes-dashboard
    helm.sh/chart: kubernetes-dashboard-5.1.1
    app.kubernetes.io/instance: RELEASE-NAME
    app.kubernetes.io/version: "2.4.0"
    app.kubernetes.io/managed-by: Helm
    app.kubernetes.io/component: kubernetes-dashboard
spec:
  replicas: 1
  strategy:
    rollingUpdate:
      maxSurge: 0
      maxUnavailable: 1
    type: RollingUpdate
  selector:
    matchLabels:

      app.kubernetes.io/name: kubernetes-dashboard
      app.kubernetes.io/instance: RELEASE-NAME
      app.kubernetes.io/component: kubernetes-dashboard
  template:
    metadata:
      annotations:
      labels:
        app.kubernetes.io/name: kubernetes-dashboard
        helm.sh/chart: kubernetes-dashboard-5.1.1
        app.kubernetes.io/instance: RELEASE-NAME
        app.kubernetes.io/version: "2.4.0"
        app.kubernetes.io/managed-by: Helm
        app.kubernetes.io/component: kubernetes-dashboard
    spec:
      securityContext:

        seccompProfile:
          type: RuntimeDefault
      serviceAccountName: RELEASE-NAME-kubernetes-dashboard
      containers:
      - name: kubernetes-dashboard
        image: "kubernetesui/dashboard:v2.4.0"
        imagePullPolicy: IfNotPresent
        args:
          - --namespace=default
          - --auto-generate-certificates
          - --sidecar-host=http://127.0.0.1:8000
        ports:
        - name: https
          containerPort: 8443
          protocol: TCP
        volumeMounts:
        - name: kubernetes-dashboard-certs
          mountPath: /certs
          # Create on-disk volume to store exec logs
        - mountPath: /tmp
          name: tmp-volume
        livenessProbe:
          httpGet:
            scheme: HTTPS
            path: /
            port: 8443
          initialDelaySeconds: 30
          timeoutSeconds: 30
        resources:

          limits:
            cpu: 2
            memory: 200Mi
          requests:
            cpu: 100m
            memory: 200Mi
        securityContext:

          allowPrivilegeEscalation: false
          readOnlyRootFilesystem: true
          runAsGroup: 2001
          runAsUser: 1001
      - name: dashboard-metrics-scraper
        image: "kubernetesui/metrics-scraper:v1.0.7"
        imagePullPolicy: IfNotPresent
        ports:
          - containerPort: 8000
            protocol: TCP
        livenessProbe:
          httpGet:
            scheme: HTTP
            path: /
            port: 8000
          initialDelaySeconds: 30
          timeoutSeconds: 30
        volumeMounts:
        - mountPath: /tmp
          name: tmp-volume
        securityContext:

          allowPrivilegeEscalation: false
          readOnlyRootFilesystem: true
          runAsGroup: 2001
          runAsUser: 1001
      volumes:
      - name: kubernetes-dashboard-certs
        secret:
          secretName: RELEASE-NAME-kubernetes-dashboard-certs
      - name: tmp-volume
        emptyDir: {}
```
**3. custom metric scraper values in custom values-test.yaml:**
```yaml
metricsScraper:
  enabled: true
  containerSecurityContext:
    allowPrivilegeEscalation: false
    readOnlyRootFilesystem: true
    runAsUser: 1002 # value for separation
    runAsGroup: 2002 # value for separation
```
Output of the Deployment:
```yaml
apiVersion: apps/v1
kind: Deployment
metadata:
  name: RELEASE-NAME-kubernetes-dashboard
  annotations:
  labels:
    app.kubernetes.io/name: kubernetes-dashboard
    helm.sh/chart: kubernetes-dashboard-5.1.1
    app.kubernetes.io/instance: RELEASE-NAME
    app.kubernetes.io/version: "2.4.0"
    app.kubernetes.io/managed-by: Helm
    app.kubernetes.io/component: kubernetes-dashboard
spec:
  replicas: 1
  strategy:
    rollingUpdate:
      maxSurge: 0
      maxUnavailable: 1
    type: RollingUpdate
  selector:
    matchLabels:

      app.kubernetes.io/name: kubernetes-dashboard
      app.kubernetes.io/instance: RELEASE-NAME
      app.kubernetes.io/component: kubernetes-dashboard
  template:
    metadata:
      annotations:
      labels:
        app.kubernetes.io/name: kubernetes-dashboard
        helm.sh/chart: kubernetes-dashboard-5.1.1
        app.kubernetes.io/instance: RELEASE-NAME
        app.kubernetes.io/version: "2.4.0"
        app.kubernetes.io/managed-by: Helm
        app.kubernetes.io/component: kubernetes-dashboard
    spec:
      securityContext:

        seccompProfile:
          type: RuntimeDefault
      serviceAccountName: RELEASE-NAME-kubernetes-dashboard
      containers:
      - name: kubernetes-dashboard
        image: "kubernetesui/dashboard:v2.4.0"
        imagePullPolicy: IfNotPresent
        args:
          - --namespace=default
          - --auto-generate-certificates
          - --sidecar-host=http://127.0.0.1:8000
        ports:
        - name: https
          containerPort: 8443
          protocol: TCP
        volumeMounts:
        - name: kubernetes-dashboard-certs
          mountPath: /certs
          # Create on-disk volume to store exec logs
        - mountPath: /tmp
          name: tmp-volume
        livenessProbe:
          httpGet:
            scheme: HTTPS
            path: /
            port: 8443
          initialDelaySeconds: 30
          timeoutSeconds: 30
        resources:

          limits:
            cpu: 2
            memory: 200Mi
          requests:
            cpu: 100m
            memory: 200Mi
        securityContext:

          allowPrivilegeEscalation: false
          readOnlyRootFilesystem: true
          runAsGroup: 2001
          runAsUser: 1001
      - name: dashboard-metrics-scraper
        image: "kubernetesui/metrics-scraper:v1.0.7"
        imagePullPolicy: IfNotPresent
        ports:
          - containerPort: 8000
            protocol: TCP
        livenessProbe:
          httpGet:
            scheme: HTTP
            path: /
            port: 8000
          initialDelaySeconds: 30
          timeoutSeconds: 30
        volumeMounts:
        - mountPath: /tmp
          name: tmp-volume
        securityContext:

          allowPrivilegeEscalation: false
          readOnlyRootFilesystem: true
          runAsGroup: 2002
          runAsUser: 1002
      volumes:
      - name: kubernetes-dashboard-certs
        secret:
          secretName: RELEASE-NAME-kubernetes-dashboard-certs
      - name: tmp-volume
        emptyDir: {}
```
**4. disable globally securityContext and containerSecurityContext in custom values-test.yaml:**
```yaml
securityContext: null
containerSecurityContext: null
```
Output of the Deployment:
```yaml
apiVersion: apps/v1
kind: Deployment
metadata:
  name: RELEASE-NAME-kubernetes-dashboard
  annotations:
  labels:
    app.kubernetes.io/name: kubernetes-dashboard
    helm.sh/chart: kubernetes-dashboard-5.1.1
    app.kubernetes.io/instance: RELEASE-NAME
    app.kubernetes.io/version: "2.4.0"
    app.kubernetes.io/managed-by: Helm
    app.kubernetes.io/component: kubernetes-dashboard
spec:
  replicas: 1
  strategy:
    rollingUpdate:
      maxSurge: 0
      maxUnavailable: 1
    type: RollingUpdate
  selector:
    matchLabels:

      app.kubernetes.io/name: kubernetes-dashboard
      app.kubernetes.io/instance: RELEASE-NAME
      app.kubernetes.io/component: kubernetes-dashboard
  template:
    metadata:
      annotations:
      labels:
        app.kubernetes.io/name: kubernetes-dashboard
        helm.sh/chart: kubernetes-dashboard-5.1.1
        app.kubernetes.io/instance: RELEASE-NAME
        app.kubernetes.io/version: "2.4.0"
        app.kubernetes.io/managed-by: Helm
        app.kubernetes.io/component: kubernetes-dashboard
    spec:
      serviceAccountName: RELEASE-NAME-kubernetes-dashboard
      containers:
      - name: kubernetes-dashboard
        image: "kubernetesui/dashboard:v2.4.0"
        imagePullPolicy: IfNotPresent
        args:
          - --namespace=default
          - --auto-generate-certificates
          - --metrics-provider=none
        ports:
        - name: https
          containerPort: 8443
          protocol: TCP
        volumeMounts:
        - name: kubernetes-dashboard-certs
          mountPath: /certs
          # Create on-disk volume to store exec logs
        - mountPath: /tmp
          name: tmp-volume
        livenessProbe:
          httpGet:
            scheme: HTTPS
            path: /
            port: 8443
          initialDelaySeconds: 30
          timeoutSeconds: 30
        resources:

          limits:
            cpu: 2
            memory: 200Mi
          requests:
            cpu: 100m
            memory: 200Mi
      volumes:
      - name: kubernetes-dashboard-certs
        secret:
          secretName: RELEASE-NAME-kubernetes-dashboard-certs
      - name: tmp-volume
        emptyDir: {}
```
**5. disable globally securityContext and containerSecurityContext but set metrics containerSecurityContext in custom values-test.yaml:**
```yaml
securityContext: null
containerSecurityContext: null
metricsScraper:
  enabled: true
  containerSecurityContext:
    allowPrivilegeEscalation: false
    readOnlyRootFilesystem: true
    runAsUser: 1002 # value for separation
    runAsGroup: 2002 # value for separation
```
Output of the Deployment:
```yaml
apiVersion: apps/v1
kind: Deployment
metadata:
  name: RELEASE-NAME-kubernetes-dashboard
  annotations:
  labels:
    app.kubernetes.io/name: kubernetes-dashboard
    helm.sh/chart: kubernetes-dashboard-5.1.1
    app.kubernetes.io/instance: RELEASE-NAME
    app.kubernetes.io/version: "2.4.0"
    app.kubernetes.io/managed-by: Helm
    app.kubernetes.io/component: kubernetes-dashboard
spec:
  replicas: 1
  strategy:
    rollingUpdate:
      maxSurge: 0
      maxUnavailable: 1
    type: RollingUpdate
  selector:
    matchLabels:

      app.kubernetes.io/name: kubernetes-dashboard
      app.kubernetes.io/instance: RELEASE-NAME
      app.kubernetes.io/component: kubernetes-dashboard
  template:
    metadata:
      annotations:
      labels:
        app.kubernetes.io/name: kubernetes-dashboard
        helm.sh/chart: kubernetes-dashboard-5.1.1
        app.kubernetes.io/instance: RELEASE-NAME
        app.kubernetes.io/version: "2.4.0"
        app.kubernetes.io/managed-by: Helm
        app.kubernetes.io/component: kubernetes-dashboard
    spec:
      serviceAccountName: RELEASE-NAME-kubernetes-dashboard
      containers:
      - name: kubernetes-dashboard
        image: "kubernetesui/dashboard:v2.4.0"
        imagePullPolicy: IfNotPresent
        args:
          - --namespace=default
          - --auto-generate-certificates
          - --sidecar-host=http://127.0.0.1:8000
        ports:
        - name: https
          containerPort: 8443
          protocol: TCP
        volumeMounts:
        - name: kubernetes-dashboard-certs
          mountPath: /certs
          # Create on-disk volume to store exec logs
        - mountPath: /tmp
          name: tmp-volume
        livenessProbe:
          httpGet:
            scheme: HTTPS
            path: /
            port: 8443
          initialDelaySeconds: 30
          timeoutSeconds: 30
        resources:

          limits:
            cpu: 2
            memory: 200Mi
          requests:
            cpu: 100m
            memory: 200Mi
      - name: dashboard-metrics-scraper
        image: "kubernetesui/metrics-scraper:v1.0.7"
        imagePullPolicy: IfNotPresent
        ports:
          - containerPort: 8000
            protocol: TCP
        livenessProbe:
          httpGet:
            scheme: HTTP
            path: /
            port: 8000
          initialDelaySeconds: 30
          timeoutSeconds: 30
        volumeMounts:
        - mountPath: /tmp
          name: tmp-volume
        securityContext:

          allowPrivilegeEscalation: false
          readOnlyRootFilesystem: true
          runAsGroup: 2002
          runAsUser: 1002
      volumes:
      - name: kubernetes-dashboard-certs
        secret:
          secretName: RELEASE-NAME-kubernetes-dashboard-certs
      - name: tmp-volume
        emptyDir: {}
```

I hope this implementation is ok for you guys :)
greez

floreks commented on aio/deploy/helm-chart/kubernetes-dashboard/templates/deployment.yaml:
```diff
@@ -143,9 +143,13 @@ spec:
           name: tmp-volume
-{{- with .Values.metricsScraper.containerSecurityContext }}
+
+{{- if .Values.metricsScraper.containerSecurityContext }}
```
Isn't it a bit weird that dashboard deployment refers to the `metricsScraper.containerSecurityContext`? It should be independent.

janlauber replied:
Sorry, I don't know exactly what you mean. This if value checks if there are separate containerSecurityContext values set for the metrics scraper. When yes, they should take place, if not the else if will check if there are containerSecurityContext globally set. This should guarantee, that people who set individual containerSecurityContext values for the metrics scraper cannot do that after upgrading to this helm chart version. 
Or do you mean there is a simpler way to achieve this behaviour in code?

floreks replied:
Sorry, I was under the impression that this is part of the "pure" kubernetes-dashboard deployment. Now that I am looking more closely I see that it is a single deployment with multiple containers. I had no idea our helm deployment differs this much from our `recommended.yaml` deployment :slightly_smiling_face: 

I think this is the best solution then.

floreks commented:
If there are no remarks from @desaintmartin then this LGTM and we can merge.

desaintmartin commented on aio/deploy/helm-chart/kubernetes-dashboard/values.yaml:
```diff
@@ -236,12 +247,13 @@ metricsScraper:
-    runAsUser: 1001
-    runAsGroup: 2001
+  ## SecurityContext especially for the kubernetes dashboard metrics scraper container
+  ## If not set, the global containterSecurityContext values will define these values
```
minor typo!

desaintmartin commented:
/lgtm
/hold cancel
let's accept this minor typo. ;)

janlauber commented:
@desaintmartin & @floreks 
Thank you very much for merging and your work!

Greez
